### PR TITLE
[JENKINS-72009, JENKINS-72200, JENKINS-24947] various fixes and improvements around disk space monitoring

### PR DESCRIPTION
**Pull request by mawinter69**
[JENKINS-24947](https://issues.jenkins.io/browse/JENKINS-24947):
- Introduce a NodeProperty to define agent specific thresholds for free disk and temp space.

[JENKINS-72009](https://issues.jenkins.io/browse/JENKINS-72009)
- Enhance the tooltip in the executor widget so it includes the threshold (also fixes a wrong conversion)
- Show a tooltip in the nodes table when agent was marked offline or has a warning about low diskspace
- Use binary prefixes when showing the disk space (GB -> GiB)
- allow to configure also a warning level
- Change text of configure button to make it clearer what it does
- Store totalsize to so we can show it in tooltips
- Show a tooltip also when all is good that contains total size and the path

[JENKINS-72200](https://issues.jenkins.io/browse/JENKINS-72200):
- ensure an agent is taken offline/online due to diskspace when the background nodemonitor runs.

an admin monitor for disk space warnings is very hard to implement, so skipping for now

support-core and metrics plugin should be adjusted to reflect agent individual thresholds.

image

See [JENKINS-72009](https://issues.jenkins.io/browse/JENKINS-72009)
[JENKINS-72200](https://issues.jenkins.io/browse/JENKINS-72200)
[JENKINS-24947](https://issues.jenkins.io/browse/JENKINS-24947)

### Testing done
- Manual testing as described in [JENKINS-72200](https://issues.jenkins.io/browse/JENKINS-72200)
- Configure node properties for agents and verify they are taken offline or show a warning

### Proposed changelog entries

- Allow configuration of disk thresholds globally and for each agent.  Improve warning when disk space is too low. Ensure agents are taken offline when disk space is low.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

N/A

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

**Comment by abhishekmaity** [comment]
@jenkinsci/core-pr-reviewers

**Review by Kevin-CB — COMMENTED** [review]
Looks fine from a security perspective

File: core/src/main/java/hudson/Functions.java
Comment (MarkEWaite) [inline core/src/main/java/hudson/Functions.java:2302]: I recommend against making this change and the related changes that correct the previously inaccurate description of the units because users are accustomed to seeing the values displayed in that format and the data entry forms for the node monitor configuration do not accept "KiB", "MiB", or "GiB".  I think it is better for users to have the displayed strings usable as the values in the input forms rather than to have the correct unit of measure displayed but not accepted by the input form.
```suggestion
            measure = "KB";
```

File: core/src/main/resources/hudson/node_monitors/DiskSpaceMonitorNodeProperty/config.jelly
Comment (MarkEWaite) [inline core/src/main/resources/hudson/node_monitors/DiskSpaceMonitorNodeProperty/config.jelly:4]: When I tried to enter a value with "GiB" suffix, a stack trace was displayed showing:
```
java.lang.NumberFormatException: For input string: "60GI"
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2054)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.base/java.lang.Double.parseDouble(Double.java:792)
	at hudson.node_monitors.DiskSpaceMonitorDescriptor$DiskSpace.parse(DiskSpaceMonitorDescriptor.java:163)
	at hudson.node_monitors.AbstractDiskSpaceMonitor.<init>(AbstractDiskSpaceMonitor.java:23)
	at hudson.node_monitors.TemporarySpaceMonitor.<init>(TemporarySpaceMonitor.java:54)
	at java.base/jdk.internal.reflect.DirectConstructorHandleAccessor.newInstance(DirectConstructorHandleAccessor.java:62)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:502)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:486)
	at org.kohsuke.stapler.RequestImpl.invokeConstructor(RequestImpl.java:604)
	at org.kohsuke.stapler.RequestImpl.instantiate(RequestImpl.java:883)
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:768)
Caused: java.lang.IllegalArgumentException: Failed to instantiate class hudson.node_monitors.TemporarySpaceMonitor from {"freeSpaceThreshold":"60GiB","freeSpaceWarningThreshold":"2GB"}
```

I think it is better to show the misleading suffix.

**Review by MarkEWaite — CHANGES_REQUESTED** [review]
I really like the direction this is moving very much.  Interactive testing found no issues.  Thanks!

File: core/src/main/resources/hudson/node_monitors/DiskSpaceMonitor/column.jelly
Comment (MarkEWaite) [inline core/src/main/resources/hudson/node_monitors/DiskSpaceMonitor/column.jelly:35]: I thought that I already added a comment with this picture, but now I can't find it.  Apologies if this is a duplicate.

The formatting is not applied to the free disk space warning in my interactive use.  It looks like this when a free disc space problem has taken the agent offline:

disk-space-monitoring-improvements

File: core/src/main/resources/hudson/node_monitors/DiskSpaceMonitorNodeProperty/config.jelly
Comment (MarkEWaite) [reply core/src/main/resources/hudson/node_monitors/DiskSpaceMonitorNodeProperty/config.jelly:4]: > from the stacktrace it looks as if you tested this not with my change. The line numbers indicate that this is the old version.

That must have been what happened.  I'll try it again later today.

File: core/src/main/resources/hudson/node_monitors/DiskSpaceMonitor/column.jelly
Comment (MarkEWaite) [reply core/src/main/resources/hudson/node_monitors/DiskSpaceMonitor/column.jelly:35]: That must be it.  Will try later today.

Comment (MarkEWaite) [reply core/src/main/resources/hudson/node_monitors/DiskSpaceMonitor/column.jelly:35]: Confirmed that it was my mistake.  Nothing wrong with pull request.  Resolving.

File: core/src/main/resources/hudson/node_monitors/DiskSpaceMonitorNodeProperty/config.jelly
Comment (MarkEWaite) [reply core/src/main/resources/hudson/node_monitors/DiskSpaceMonitorNodeProperty/config.jelly:4]: Confirmed that you are correct @mawinter69 .  I was using a half-baked configuration because I had failed to rebuild after switching context from one pull request to the next.  Resolving

File: core/src/main/java/hudson/Functions.java
Comment (MarkEWaite) [reply core/src/main/java/hudson/Functions.java:2302]: New units worked for me as well.  As far as I can tell, the old units are still accepted.  I think that allowing both old and new units in the new implementation is enough.  We don't promise that new settings can be taken backwards to old releases.

The configuration as code settings are especially nice.  It let's me handle some of my static agents as the special cases that they are, while still managing them with configuration as code.  Agents with tiny disc drives will be able to have smaller thresholds while agents with reasonable drives can user larger thresholds for a better margin of safety.  Thanks very much!

**Comment by MarkEWaite** [comment]
Interactive testing showed that all the settings worked very well.  I confirmed that the settings are visible through configuration as code and that agents are taken offline when the thresholds are exceeded.

**Review by MarkEWaite — APPROVED** [review]
Thanks very much!

**Review by timja — APPROVED** [review]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/hudson/node_monitors/AbstractDiskSpaceMonitor.java
Comment (jglick) [inline core/src/main/java/hudson/node_monitors/AbstractDiskSpaceMonitor.java:29]: Note: this is one of the unusual cases where a settings format change is not forward-compatible. So if you run a controller from a version after this PR and then downgrade to a version prior to it and try to save monitor config, you will get an error

```
java.lang.NumberFormatException: For input string: "1GI"
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2054)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.base/java.lang.Double.parseDouble(Double.java:792)
	at hudson.node_monitors.DiskSpaceMonitorDescriptor$DiskSpace.parse(DiskSpaceMonitorDescriptor.java:163)
	at hudson.node_monitors.AbstractDiskSpaceMonitor.<init>(AbstractDiskSpaceMonitor.java:23)
	at hudson.node_monitors.DiskSpaceMonitor.<init>(DiskSpaceMonitor.java:49)
	at …
Caused: java.lang.IllegalArgumentException: Failed to instantiate class hudson.node_monitors.DiskSpaceMonitor from {"freeSpaceThreshold":"1GiB"}
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:771)
	at org.kohsuke.stapler.RequestImpl.bindJSON(RequestImpl.java:551)
	at org.kohsuke.stapler.RequestImpl.bindJSON(RequestImpl.java:546)
	at hudson.model.Descriptor.bindJSON(Descriptor.java:623)
	at hudson.model.Descriptor.newInstance(Descriptor.java:593)
Caused: java.lang.LinkageError: Failed to instantiate class hudson.node_monitors.DiskSpaceMonitor from {"freeSpaceThreshold":"1GiB"}
	at hudson.model.Descriptor.newInstance(Descriptor.java:596)
	at hudson.util.DescribableList.rebuild(DescribableList.java:183)
	at hudson.model.ComputerSet.doConfigSubmit(ComputerSet.java:355)
	at …
```

**Pull request merged**
